Feature/eval with taxonomy

lessej opened the pull request:
**Context**

For models that have a taxonomy field, this allows for running an object level analysis using the taxonomic structure.

It first

- Builds a list of labels and their taxonomic children using an intersection between the project labels and the ML model's categories
- Gathers a list of images based on the project and start and end dates
- Analyzes each object level label using the taxonomic child information

This is mostly adapted from the `analyzeMLObjectLevel.js` script

nathanielrindlaub commented on src/scripts/analyzeMLObjectLevelWithTaxonomy.js:
```diff
@@ -0,0 +1,376 @@
+          // (i.e., for "rodent" prediction, a firstValidLabel of "rodent" or any of its taxonomic children),
+          if (
+            obj.locked &&
+            objectLabelsHaveTargetLabel &&
```
It appears as though we are undercounting true positives for higher-level target classes (classes that have children). For example, in the palmyra bucket test data, for the target class of `muridae family:9880b662-dc21-453b-aa2f-dd97338f623b` with validation classes of `brown rat, ff803ff7-ed2b-4520-9304-99b408657f47, rattus species, 00049ff0-2ffa-4d82-8cf3-c861fbbfa9d5, muridae family, 9880b662-dc21-453b-aa2f-dd97338f623b`, we have 53 `actuals` but only 1 `truePositive`.

At a glance I think it's because the `objectLabelsHaveTargetLabel` check will only count objects that have a ml prediction of the target class towards true positives, but we should also be including objects that have an ml prediction of all of the child classes as well (`brown rat`, `rattus species`). If we're evaluating the model's `muridae family` class performance, we should count predictions of `muridae family` children as correct.

nathanielrindlaub replied:
Ok I updated it but copying the definitons here too. Might want to sanity check these to make sure they make sense to you too:
- an `actual` of a given predicted class would be any object who's FVL is an exact match or child taxa of the class, regardless of whether the predicted class was actually predicted for that object or not.
- a `truePositive` of a given predicted class would be any object that both (a) received a prediction of that class (or child taxa of the class) and (b) who's FVL is an exact match or child taxa of the predicted class.
- a `falsePositive` of a given predicted class would be any object that both (a) received a prediction of that class (or child taxa of the class) and (b) who's FVL is NOT an exact match nor child taxa of the predicted class.
- a `falseNegative` of a given predicted class would be any object that (a) didn't receive a prediction of that class (or any child taxa of that class) but (b) who's FVL is an exact match or child taxa of the predicted class.

nathanielrindlaub commented on src/scripts/analyzeMLObjectLevelWithTaxonomy.js:
```diff
@@ -0,0 +1,376 @@
+  }
+
+  return tree;
+};
```
On second thought, maybe I'm just getting thrown off by this being named `tree` and assuming "tree" refers to taxonomic tree. Really what you're building is a list of predicted labels and their respective validating labels. Also when I think "tree" i think of a multi-tiered graph representation but what we're building here a pretty flat structure. Maybe we can call it `buildValidatingLabelList()` or something like that?

I do think sticking with the `"<labelName>:<labelId>"` naming convention would be helpful. So instead of the output of the `getMLLabels()` function below for `"rodent"` being:

```javascript
  {
    "name": "rodent",
    "labelId": "90d950db-2106-4bd9-a4c1-777604c3eada",
    "taxonomicChildren": [
      "american beaver",
      "43320a08-bf31-49a5-8213-f032311c5765",
      "brown rat",
      "ff803ff7-ed2b-4520-9304-99b408657f47",
      "rattus species",
      "00049ff0-2ffa-4d82-8cf3-c861fbbfa9d5",
      "muridae family",
      "9880b662-dc21-453b-aa2f-dd97338f623b",
      "rodent",
      "90d950db-2106-4bd9-a4c1-777604c3eada"
    ]
  }
```

I think the following is more compact and accurate: 

```
  {
    "predicted": "rodent:90d950db-2106-4bd9-a4c1-777604c3eada",
    "validating": [
      "american beaver:43320a08-bf31-49a5-8213-f032311c5765",
      "brown rat:ff803ff7-ed2b-4520-9304-99b408657f47",
      "rattus species:00049ff0-2ffa-4d82-8cf3-c861fbbfa9d5",
      "muridae family:9880b662-dc21-453b-aa2f-dd97338f623b",
      "rodent:90d950db-2106-4bd9-a4c1-777604c3eada"
    ]
  }
```

lessej commented on src/scripts/analyzeMLObjectLevelWithTaxonomy.js:
```diff
@@ -0,0 +1,376 @@
+          const objectLabelsHaveTargetLabel = obj.labels.some((label) => (
+            label.type === 'ml' &&
+            label.mlModel === ML_MODEL &&
+            (label.labelId === mlLabel.labelId || label.labelId === mlLabel.name)
```
This should also look for taxonomic children (valid predicted classes)

lessej commented:
**Closed by accident

@nathanielrindlaub fixed up the things we talked about.  In particular, I went through and renamed functions / variables to be a little bit more intuitive, updated the structure of the intermediate predicted and validation objects, and updated the check used for analysis to include both the target class and the validations for the target class.  Here's a sample of the output data.  The recall does seem to go up with the hierarchy level which seems good!

[palmyra_bucket_test_speciesnet-classifier_2023-4-28--2025-10-17_object-level_2025-10-18T0032Z_config.json](https://github.com/user-attachments/files/22981347/palmyra_bucket_test_speciesnet-classifier_2023-4-28--2025-10-17_object-level_2025-10-18T0032Z_config.json)
[palmyra_bucket_test_speciesnet-classifier_2023-4-28--2025-10-17_object-level_2025-10-18T0032Z.csv](https://github.com/user-attachments/files/22981348/palmyra_bucket_test_speciesnet-classifier_2023-4-28--2025-10-17_object-level_2025-10-18T0032Z.csv)

nathanielrindlaub commented on src/scripts/analyzeMLObjectLevelWithTaxonomy.js:
```diff
@@ -0,0 +1,385 @@
+          }
+
+          // FALSE POSITIVE - object must be:
+          // (a) locked, (b) has an ml-predicted label of the target class, and
```
same note as above.

nathanielrindlaub commented on src/scripts/analyzeMLObjectLevelWithTaxonomy.js:
```diff
@@ -0,0 +1,385 @@
+          // FALSE NEGATIVE - object must be:
+          // (a) locked, (b) has a first valid label that validates the prediction/target class,
+          // (i.e., for "rodent" prediction, a firstValidLabel of "rodent" or any of its taxonomic children),
+          // and (c) does NOT have an ml-predicted label of the target class
```
" ...(c) does NOT have an ml-predicted label of the target class or any of its taxonomic children."

nathanielrindlaub reviewed (commented):
@jesse this is fantastic. I have one final flip flop on variable naming/structure, but feel free to take it or leave it - this looks great to me and I'm happy to merge it as is. 

The next step will be to make this logic work for sequence-level analysis, which I'm stoked to see the results of. But we can start that in a separate PR.
